File: backend/services/recommendation_parser.py
"""Service to parse BUY/SELL/HOLD recommendations from reports into structured schema."""

import re
from typing import Optional
from pathlib import Path

from models.schemas import ParsedRecommendation


class RecommendationParser:
    """Parse recommendations from report files; returns Pydantic ParsedRecommendation."""

    @staticmethod
    def parse_recommendation(report_content: str) -> Optional[ParsedRecommendation]:
        """
        Parse recommendation from report content.

        Returns:
            ParsedRecommendation with recommendation (BUY/SELL/HOLD), confidence, source,
            or None if no recommendation found.
        """
        if not report_content:
            return None

        # Try final_trade_decision pattern first
        pattern1 = r'\*\*Recommendation:\s*(Buy|Sell|Hold)\*\*'
        match = re.search(pattern1, report_content, re.IGNORECASE)
        if match:
            return ParsedRecommendation(
                recommendation=match.group(1).upper(),
                confidence=1.0,
                source="final_trade_decision"
            )

        # Try trader_investment_plan pattern
        pattern2 = r'FINAL TRANSACTION PROPOSAL:\s*\*\*(Buy|Sell|Hold)\*\*'
        match = re.search(pattern2, report_content, re.IGNORECASE)
        if match:
            return ParsedRecommendation(
                recommendation=match.group(1).upper(),
                confidence=0.9,
                source="trader_investment_plan"
            )

        # Try alternative patterns
        patterns = [
            r'Recommendation:\s*(Buy|Sell|Hold)',
            r'recommendation:\s*["\']?(Buy|Sell|Hold)["\']?',  # YAML frontmatter
            r'decision to\s*\*\*(Buy|Sell|Hold)\*\*',  # "the decision to **Buy** is prescribed"
            r'recommendation is to\s*(buy|sell|hold)',
            r'recommend\s*(buying|selling|holding)',
        ]
        for pattern in patterns:
            match = re.search(pattern, report_content, re.IGNORECASE)
            if match:
                rec_text = match.group(1).upper()
                if 'BUY' in rec_text or rec_text == 'BUYING':
                    rec = 'BUY'
                elif 'SELL' in rec_text or rec_text == 'SELLING':
                    rec = 'SELL'
                elif 'HOLD' in rec_text or rec_text == 'HOLDING':
                    rec = 'HOLD'
                else:
                    continue
                return ParsedRecommendation(
                    recommendation=rec,
                    confidence=0.7,
                    source="general_parsing"
                )

        return None

    @staticmethod
    def get_recommendation_from_file(file_path: Path) -> Optional[ParsedRecommendation]:
        """Get recommendation from a report file."""
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                content = f.read()
            return RecommendationParser.parse_recommendation(content)
        except Exception:
            return None

File: backend-node/src/services/recommendationParser.ts
/**
 * Parse BUY/SELL/HOLD from report content. Port of backend/services/recommendation_parser.py
 */

export interface ParsedRecommendation {
  recommendation: "BUY" | "SELL" | "HOLD";
  confidence: number | null;
  source: string;
}

export function parseRecommendation(reportContent: string | null | undefined): ParsedRecommendation | null {
  if (!reportContent?.trim()) return null;

  const content = reportContent;

  let m = content.match(/\*\*Recommendation:\s*(Buy|Sell|Hold)\*\*/i);
  if (m) {
    return {
      recommendation: m[1].toUpperCase() as "BUY" | "SELL" | "HOLD",
      confidence: 1.0,
      source: "final_trade_decision",
    };
  }

  m = content.match(/FINAL TRANSACTION PROPOSAL:\s*\*\*(Buy|Sell|Hold)\*\*/i);
  if (m) {
    return {
      recommendation: m[1].toUpperCase() as "BUY" | "SELL" | "HOLD",
      confidence: 0.9,
      source: "trader_investment_plan",
    };
  }

  const patterns: RegExp[] = [
    /Recommendation:\s*(Buy|Sell|Hold)/i,
    /recommendation:\s*["']?(Buy|Sell|Hold)["']?/i,
    /decision to\s*\*\*(Buy|Sell|Hold)\*\*/i,
    /recommendation is to\s*(buy|sell|hold)/i,
    /recommend\s*(buying|selling|holding)/i,
  ];

  for (const pattern of patterns) {
    m = content.match(pattern);
    if (m) {
      const recText = m[1].toUpperCase();
      let rec: "BUY" | "SELL" | "HOLD";
      if (recText.includes("BUY") || recText === "BUYING") rec = "BUY";
      else if (recText.includes("SELL") || recText === "SELLING") rec = "SELL";
      else if (recText.includes("HOLD") || recText === "HOLDING") rec = "HOLD";
      else continue;
      return { recommendation: rec, confidence: 0.7, source: "general_parsing" };
    }
  }

  return null;
}
